File: Wifi Range [19 - April - 2023]/Approach 2.java
//   >> JAVA CODE <<


boolean wifiRange(int N, String S, int X) 
    { 
        int[] visited = new int[N];
        for (int i = 0; i < N; i++) {
            if (S.charAt(i) == '1') {
                int start = Math.max(i - X, 0);
                int end =i + X;
                visited[start] += 1;
                if (end + 1 < N) {
                    visited[end + 1] += -1;
                }
            }
        }
        
        for(int i=1;i<N;i++) visited[i]+=visited[i-1];
        
        
        for (int val : visited) {
            if (val==0) {
                return false;
            }
        }
        
        return true;
    }

File: Wifi Range [19 - April - 2023]/Approach 2.cpp
//    >> C++ CODE <<

bool wifiRange(int N, string S, int X){
        vector<int> visited(N, 0);
        for (int i = 0; i < N; i++) {
            if (S[i] == '1') {
                visited[max(i - X, 0)] += 1;
                if (i + X + 1 < N) {
                    visited[i + X + 1] += -1;
                }
            }
        }
        for (int i = 1; i < N; i++) {
            visited[i] += visited[i - 1];
        }
        for (int val : visited) {
            if (val == 0) {
                return false;
            }
        }
        return true;
        
    }